### PR TITLE
Test Enhancement for Tcpdump test

### PR DESCRIPTION
**Pull request by smruti77**
included three more test for capturing packets in tcpdump.

Signed-off-by: bismurti bidhibrata pattajoshi <bbidhibr@in.ibm.com>

### CODE REVIEW

**Comment by smruti77** [comment]
JOB ID     : 9fbed1738559e2ca8a41882d6abbe9d66630e04c
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2019-09-18T15.25-9fbed17/job.log
 (1/4) tcpdump.py:TcpdumpTest.test;run-c7a8: PASS (103.07 s)
 (2/4) tcpdump.py:TcpdumpTest.test_protocal;run-c7a8: PASS (103.08 s)
 (3/4) tcpdump.py:TcpdumpTest.test_host;run-c7a8: PASS (103.08 s)
 (4/4) tcpdump.py:TcpdumpTest.test_src;run-c7a8: PASS (104.08 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 413.53 s
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2019-09-18T15.25-9fbed17/results.html

File: io/net/tcpdump.py [hunk 1]
```diff
@@ -50,25 +52,66 @@ def setUp(self):
...
+    def test(self):
+        """
+        Performs the tcpdump test.
+        """
+        output_file = os.path.join(self.outputdir, 'tcpdump')
+        cmd = "tcpdump -i %s -n -c %s -w '%s'" % (self.iface, self.count,
+                                                  output_file)
+        self.get_cmd_result(cmd)
+
+    def test_protocal(self):
```
Comment (narasimhan-v) [inline io/net/tcpdump.py]: @smruti77 these can be handled via different yaml options, no need of different functions.

Also, -w option is needed for all options.

**Comment by smruti77** [comment]
JOB ID     : 76b5432a1b519526dab21e8a38f170a3643c0018
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2019-09-24T10.36-76b5432/job.log
 (1/7) tcpdump.py:TcpdumpTest.test;generic-6904: PASS (166.26 s)
 (2/7) tcpdump.py:TcpdumpTest.test;generic_host-0e78: PASS (127.02 s)
 (3/7) tcpdump.py:TcpdumpTest.test;generic_src-4386: PASS (162.33 s)
 (4/7) tcpdump.py:TcpdumpTest.test;generic_dst-4d02: PASS (177.27 s)
 (5/7) tcpdump.py:TcpdumpTest.test;potocol_tcp-74da: PASS (107.54 s)
 (6/7) tcpdump.py:TcpdumpTest.test;potocol_udp-5dd3: PASS (169.17 s)
 (7/7) tcpdump.py:TcpdumpTest.test;potocol_icmp-7594: PASS (69.61 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1057.78 s
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2019-09-24T10.36-76b5432/results.html

File: io/net/tcpdump.py [hunk 2]
```diff
@@ -59,17 +82,42 @@ def test(self):
...
         obj.start()
         output_file = os.path.join(self.outputdir, 'tcpdump')
-        cmd = "tcpdump -i %s -n -c %s -w '%s'" % (self.iface, self.count,
-                                                  output_file)
+        cmd = "tcpdump -i %s -n -c %s" % (self.iface, self.count)
+        if self.option in ('host', 'src'):
+            cmd = "%s %s %s" % (cmd, self.option, self.host_ip)
+        elif self.option == "dst":
+            cmd = "%s %s %s" % (cmd, self.option, self.peer_ip)
+        elif self.option == "tcp":
```
Comment (narasimhan-v) [inline io/net/tcpdump.py]:
```
elif self.option in ['tcp', 'udp', 'icmp']:
    obj = self.nping(self.option)
    obj.start()
    cmd = "%s %s" % (cmd, self.option)
else:
    cmd = "%s %s" % (cmd, self.option)
```

File: io/net/tcpdump.py [hunk 3]
```diff
@@ -59,17 +82,42 @@ def test(self):
...
+            obj.start()
+            cmd = "%s %s" % (cmd, self.option)
+        else:
+            cmd = "%s %s" % (cmd, self.option)
+        cmd = "%s -w '%s'" % (cmd, output_file)
         for line in process.run(cmd, shell=True,
                                 ignore_status=True).stderr.decode("utf-8") \
                                                    .splitlines():
-            if "packets dropped by interface" in line:
+            if "packets dropped by kernel" in line:
```
Comment (narasimhan-v) [inline io/net/tcpdump.py]: what is the reason this is changed ?

File: io/net/tcpdump.py.data/tcpdump.yaml [hunk 4]
```diff
@@ -1,5 +1,24 @@
 interface: net0
 peer_ip:
 count: 100
+host_ip:
```
Comment (narasimhan-v) [inline io/net/tcpdump.py.data/tcpdump.yaml]: can't we get host IP from the interface ?

File: io/net/tcpdump.py (hunk 3 above)
Comment (smruti77) [reply io/net/tcpdump.py]: beacuse after running command not getting, packets dropped by interface.i am getting like this,so i changed.
10 packets captured
10 packets received by filter
0 packets dropped by kernel

Comment (narasimhan-v) [reply io/net/tcpdump.py]: can you run for more than 500 packets and check please ?

Comment (narasimhan-v) [reply io/net/tcpdump.py]: Packets dropped by kernel might include other interfaces also. We are interested only in packets dropped by one interface under test.

File: io/net/tcpdump.py.data/tcpdump.yaml (hunk 4 above)
Comment (narasimhan-v) [reply io/net/tcpdump.py.data/tcpdump.yaml]: What is the solution here ?
Also, I see duplicate `host_ip` here

File: io/net/tcpdump.py [hunk 5]
```diff
@@ -63,8 +86,18 @@ def test(self):
...
         output_file = os.path.join(self.outputdir, 'tcpdump')
-        cmd = "tcpdump -i %s -n -c %s -w '%s'" % (self.iface, self.count,
-                                                  output_file)
+        cmd = "tcpdump -i %s -n -c %s" % (self.iface, self.count)
+        if self.option in ('host', 'src'):
+            cmd = "%s %s %s" % (cmd, self.option, self.host_ip)
+        elif self.option == "dst":
+            cmd = "%s %s %s" % (cmd, self.option, self.peer_ip)
+        elif self.option in ('tcp', 'udp', 'icmp'):
+            obj = self.nping(self.option)
```
Comment (narasimhan-v) [inline io/net/tcpdump.py]: replacing the `obj` in line 86 ?

File: io/net/tcpdump.py [hunk 6]
```diff
@@ -63,8 +86,18 @@ def test(self):
...
+            obj.start()
+            cmd = "%s %s" % (cmd, self.option)
+        else:
+            cmd = "%s %s" % (cmd, self.option)
+        cmd = "%s -w '%s'" % (cmd, output_file)
```
Comment (narasimhan-v) [inline io/net/tcpdump.py]: This can be enhanced further to have `cmd` defined as a generic variable and then options added based on yaml parameters.

**Comment by smruti77** [comment]
@narasimhan-v  same thing only i did here, but last i added another cmd its because of we need to write in to the file after that.

**Comment by smruti77** [comment]
JOB ID     : b581bea897f49519f8780bc16dd2d04abb636e6f
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2019-10-17T11.26-b581bea/job.log
 (1/7) tcpdump.py:TcpdumpTest.test;generic-d121: PASS (46.73 s)
 (2/7) tcpdump.py:TcpdumpTest.test;generic_host-36e3: PASS (34.46 s)
 (3/7) tcpdump.py:TcpdumpTest.test;generic_src-3257: PASS (37.31 s)
 (4/7) tcpdump.py:TcpdumpTest.test;generic_dst-b0d9: PASS (37.26 s)
 (5/7) tcpdump.py:TcpdumpTest.test;protocol_tcp-52a2: PASS (34.32 s)
 (6/7) tcpdump.py:TcpdumpTest.test;protocol_udp-adc3: PASS (35.46 s)
 (7/7) tcpdump.py:TcpdumpTest.test;protocol_icmp-7ebe: PASS (34.28 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 260.21 s
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2019-10-17T11.26-b581bea/results.html

**Comment by smruti77** [comment]
[job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/3738116/job.log)

**Comment by narasimhan-v** [comment]
@vaishnavibhat review please

File: io/net/tcpdump.py [hunk 7]
```diff
@@ -52,19 +57,47 @@ def setUp(self):
 
         # Install needed packages
         smm = SoftwareManager()
-        if not smm.check_installed("tcpdump") and not smm.install("tcpdump"):
-            self.cancel("Can not install tcpdump")
+        detected_distro = distro.detect()
+        pkgs = ['tcpdump']
```
Comment (vaishnavibhat) [inline io/net/tcpdump.py]: Please check and nmap package for RHEL distro. Few versions like RH7, does not have nmap installed by default.

**Comment by smruti77** [comment]
@vaishnavibhat only for rhel 7 ..?

**Comment by smruti77** [comment]
JOB ID     : 6d6cb7c27a995ab7943ebffc9fbc86a79aee0f36
JOB LOG    : /root/avocado-fvt-wrapper/results/job-2019-10-21T11.27-6d6cb7c/job.log
 (1/7) tcpdump.py:TcpdumpTest.test;generic-5307: PASS (48.70 s)
 (2/7) tcpdump.py:TcpdumpTest.test;generic_host-4158: PASS (36.36 s)
 (3/7) tcpdump.py:TcpdumpTest.test;generic_src-abce: PASS (38.38 s)
 (4/7) tcpdump.py:TcpdumpTest.test;generic_dst-09b7: PASS (37.64 s)
 (5/7) tcpdump.py:TcpdumpTest.test;protocol_tcp-3f5d: PASS (36.05 s)
 (6/7) tcpdump.py:TcpdumpTest.test;protocol_udp-6a7a: PASS (36.35 s)
 (7/7) tcpdump.py:TcpdumpTest.test;protocol_icmp-118b: PASS (36.12 s)
RESULTS    : PASS 7 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 269.94 s
JOB HTML   : /root/avocado-fvt-wrapper/results/job-2019-10-21T11.27-6d6cb7c/results.html

**Pull request merged**